File: app/src/main/java/com/aar/app/wordsearch/features/gameplay/GamePlayActivity.java
package com.aar.app.wordsearch.features.gameplay;

import android.animation.Animator;
import android.animation.AnimatorInflater;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.Paint;
import android.os.Bundle;
import android.os.Handler;
import android.util.DisplayMetrics;
import android.view.View;
import android.widget.TextView;

import com.aar.app.wordsearch.R;
import com.aar.app.wordsearch.features.SoundPlayer;
import com.aar.app.wordsearch.features.ViewModelFactory;
import com.aar.app.wordsearch.WordSearchApp;
import com.aar.app.wordsearch.commons.DurationFormatter;
import com.aar.app.wordsearch.commons.Util;
import com.aar.app.wordsearch.model.GameData;
import com.aar.app.wordsearch.custom.LetterBoard;
import com.aar.app.wordsearch.custom.StreakView;
import com.aar.app.wordsearch.custom.layout.FlowLayout;
import com.aar.app.wordsearch.features.gameover.GameOverActivity;
import com.aar.app.wordsearch.features.FullscreenActivity;
import com.aar.app.wordsearch.model.UsedWord;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindColor;
import butterknife.BindView;
import butterknife.ButterKnife;

public class GamePlayActivity extends FullscreenActivity {

    public static final String EXTRA_GAME_ROUND_ID =
            "com.aar.app.wordsearch.features.gameplay.GamePlayActivity.ID";
    public static final String EXTRA_ROW_COUNT =
            "com.aar.app.wordsearch.features.gameplay.GamePlayActivity.ROW";
    public static final String EXTRA_COL_COUNT =
            "com.aar.app.wordsearch.features.gameplay.GamePlayActivity.COL";

    private static final StreakLineMapper STREAK_LINE_MAPPER = new StreakLineMapper();

    @Inject
    SoundPlayer mSoundPlayer;

    @Inject ViewModelFactory mViewModelFactory;
    private GamePlayViewModel mViewModel;

    @BindView(R.id.text_duration) TextView mTextDuration;
    @BindView(R.id.letter_board) LetterBoard mLetterBoard;
    @BindView(R.id.flow_layout) FlowLayout mFlowLayout;

    @BindView(R.id.text_sel_layout) View mTextSelLayout;
    @BindView(R.id.text_selection) TextView mTextSelection;

    @BindView(R.id.answered_text_count) TextView mAnsweredTextCount;
    @BindView(R.id.words_count) TextView mWordsCount;

    @BindView(R.id.finished_text) TextView mFinishedText;

    @BindView(R.id.loading) View mLoading;
    @BindView(R.id.loadingText) TextView mLoadingText;
    @BindView(R.id.content_layout) View mContentLayout;

    @BindColor(R.color.gray) int mGrayColor;

    private ArrayLetterGridDataAdapter mLetterAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_game_play);

        ButterKnife.bind(this);
        ((WordSearchApp) getApplication()).getAppComponent().inject(this);

        mLetterBoard.getStreakView().setEnableOverrideStreakLineColor(getPreferences().grayscale());
        mLetterBoard.getStreakView().setOverrideStreakLineColor(mGrayColor);
        mLetterBoard.setOnLetterSelectionListener(new LetterBoard.OnLetterSelectionListener() {
            @Override
            public void onSelectionBegin(StreakView.StreakLine streakLine, String str) {
                streakLine.setColor(Util.getRandomColorWithAlpha(170));
                mTextSelLayout.setVisibility(View.VISIBLE);
                mTextSelection.setText(str);
            }

            @Override
            public void onSelectionDrag(StreakView.StreakLine streakLine, String str) {
                if (str.isEmpty()) {
                    mTextSelection.setText("...");
                } else {
                    mTextSelection.setText(str);
                }
            }

            @Override
            public void onSelectionEnd(StreakView.StreakLine streakLine, String str) {
                mViewModel.answerWord(str, STREAK_LINE_MAPPER.revMap(streakLine), getPreferences().reverseMatching());
                mTextSelLayout.setVisibility(View.GONE);
                mTextSelection.setText(str);
            }
        });

        mViewModel = ViewModelProviders.of(this, mViewModelFactory).get(GamePlayViewModel.class);
        mViewModel.getOnTimer().observe(this, this::showDuration);
        mViewModel.getOnGameState().observe(this, this::onGameStateChanged);
        mViewModel.getOnAnswerResult().observe(this, this::onAnswerResult);

        Bundle extras = getIntent().getExtras();
        if (extras != null) {
            if (extras.containsKey(EXTRA_GAME_ROUND_ID)) {
                int gid = extras.getInt(EXTRA_GAME_ROUND_ID);
                mViewModel.loadGameRound(gid);
            } else {
                int rowCount = extras.getInt(EXTRA_ROW_COUNT);
                int colCount = extras.getInt(EXTRA_COL_COUNT);
                mViewModel.generateNewGameRound(rowCount, colCount);
            }
        }

        if (!getPreferences().showGridLine()) {
            mLetterBoard.getGridLineBackground().setVisibility(View.INVISIBLE);
        } else {
            mLetterBoard.getGridLineBackground().setVisibility(View.VISIBLE);
        }

        mLetterBoard.getStreakView().setSnapToGrid(getPreferences().getSnapToGrid());
        mFinishedText.setVisibility(View.GONE);
    }

    @Override
    protected void onStart() {
        super.onStart();
        mViewModel.resumeGame();
    }

    @Override
    protected void onStop() {
        super.onStop();
        mViewModel.pauseGame();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mViewModel.stopGame();
    }

    private void onAnswerResult(GamePlayViewModel.AnswerResult answerResult) {
        if (answerResult.correct) {
            TextView textView = findUsedWordTextViewByUsedWordId(answerResult.usedWordId);
            if (textView != null) {
                UsedWord uw = (UsedWord) textView.getTag();

                if (getPreferences().grayscale()) {
                    uw.getAnswerLine().color = mGrayColor;
                }
                textView.setBackgroundColor(uw.getAnswerLine().color);
                textView.setText(uw.getString());
                textView.setTextColor(Color.WHITE);
                textView.setPaintFlags(textView.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);

                Animator anim = AnimatorInflater.loadAnimator(this, R.animator.zoom_in_out);
                anim.setTarget(textView);
                anim.start();
            }

            mSoundPlayer.play(SoundPlayer.Sound.Correct);
        }
        else {
            mLetterBoard.popStreakLine();

            mSoundPlayer.play(SoundPlayer.Sound.Wrong);
        }
    }

    private void onGameStateChanged(GamePlayViewModel.GameState gameState) {
        showLoading(false, null);
        if (gameState instanceof GamePlayViewModel.Generating) {
            GamePlayViewModel.Generating state = (GamePlayViewModel.Generating) gameState;
            String text = "Generating " + state.rowCount + "x" + state.colCount + " grid";
            showLoading(true, text);
        } else if (gameState instanceof GamePlayViewModel.Finished) {
            showFinishGame(((GamePlayViewModel.Finished) gameState).mGameData.getId());
        } else if (gameState instanceof GamePlayViewModel.Paused) {

        } else if (gameState instanceof GamePlayViewModel.Playing) {
            onGameRoundLoaded(((GamePlayViewModel.Playing) gameState).mGameData);
        }
    }

    private void onGameRoundLoaded(GameData gameData) {
        if (gameData.isFinished()) {
            setGameAsAlreadyFinished();
        }

        showLetterGrid(gameData.getGrid().getArray());
        showDuration(gameData.getDuration());
        showUsedWords(gameData.getUsedWords());
        showWordsCount(gameData.getUsedWords().size());
        showAnsweredWordsCount(gameData.getAnsweredWordsCount());
        doneLoadingContent();
    }

    private void tryScale() {
        DisplayMetrics metrics = new DisplayMetrics();
        getWindowManager().getDefaultDisplay().getMetrics(metrics);

        int boardWidth = mLetterBoard.getWidth();
        int screenWidth = metrics.widthPixels;

        if (getPreferences().autoScaleGrid() || boardWidth > screenWidth) {
            float scale = (float)screenWidth / (float)boardWidth;
            mLetterBoard.scale(scale, scale);
//            mLetterBoard.animate()
//                    .scaleX(scale)
//                    .scaleY(scale)
//                    .setDuration(400)
//                    .setInterpolator(new DecelerateInterpolator())
//                    .start();
        }
    }

    private void doneLoadingContent() {
        // call tryScale() on the next render frame
        new Handler().postDelayed(this::tryScale, 100);
    }

    private void showLoading(boolean enable, String text) {
        if (enable) {
            mLoading.setVisibility(View.VISIBLE);
            mLoadingText.setVisibility(View.VISIBLE);
            mContentLayout.setVisibility(View.GONE);
            mLoadingText.setText(text);
        } else {
            mLoading.setVisibility(View.GONE);
            mLoadingText.setVisibility(View.GONE);
            mContentLayout.setVisibility(View.VISIBLE);
        }
    }

    private void showLetterGrid(char[][] grid) {
        if (mLetterAdapter == null) {
            mLetterAdapter = new ArrayLetterGridDataAdapter(grid);
            mLetterBoard.setDataAdapter(mLetterAdapter);
        }
        else {
            mLetterAdapter.setGrid(grid);
        }
    }

    private void showDuration(int duration) {
        mTextDuration.setText(DurationFormatter.fromInteger(duration));
    }

    private void showUsedWords(List<UsedWord> usedWords) {
        for (UsedWord uw : usedWords) {
            mFlowLayout.addView( createUsedWordTextView(uw) );
        }
    }

    private void showAnsweredWordsCount(int count) {
        mAnsweredTextCount.setText(String.valueOf(count));
    }

    private void showWordsCount(int count) {
        mWordsCount.setText(String.valueOf(count));
    }

    private void showFinishGame(int gameId) {
        Intent intent = new Intent(this, GameOverActivity.class);
        intent.putExtra(GameOverActivity.EXTRA_GAME_ROUND_ID, gameId);
        startActivity(intent);
        finish();
    }

    private void setGameAsAlreadyFinished() {
        mLetterBoard.getStreakView().setInteractive(false);
        mFinishedText.setVisibility(View.VISIBLE);
    }

    //
    private TextView createUsedWordTextView(UsedWord uw) {
        TextView tv = new TextView(this);
        tv.setPadding(10, 5, 10, 5);
        if (uw.isAnswered()) {
            if (getPreferences().grayscale()) {
                uw.getAnswerLine().color = mGrayColor;
            }
            tv.setBackgroundColor(uw.getAnswerLine().color);
            tv.setText(uw.getString());
            tv.setTextColor(Color.WHITE);
            tv.setPaintFlags(tv.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);

            mLetterBoard.addStreakLine(STREAK_LINE_MAPPER.map(uw.getAnswerLine()));
        }
        else {
            String str = uw.getString();
            if (uw.isMystery()) {
                int revealCount = uw.getRevealCount();
                String uwString = uw.getString();
                str = "";
                for (int i = 0; i < uwString.length(); i++) {
                    if (revealCount > 0) {
                        str += uwString.charAt(i);
                        revealCount--;
                    }
                    else {
                        str += " ?";
                    }
                }
            }
            tv.setText(str);
        }

        tv.setTag(uw);
        return tv;
    }

    private TextView findUsedWordTextViewByUsedWordId(int usedWordId) {
        for (int i = 0; i < mFlowLayout.getChildCount(); i++) {
            TextView tv = (TextView) mFlowLayout.getChildAt(i);
            UsedWord uw = (UsedWord) tv.getTag();
            if (uw != null && uw.getId() == usedWordId) {
                return tv;
            }
        }

        return null;
    }
}

File: app/src/main/java/co/africanwolf/hiddenwords/gameplay/GamePlayActivity.kt
package co.africanwolf.hiddenwords.gameplay

import android.animation.AnimatorInflater
import android.content.Intent
import android.graphics.Color
import android.graphics.Paint
import android.os.Bundle
import android.os.Handler
import android.util.DisplayMetrics
import android.util.TypedValue
import android.view.View
import android.view.View.GONE
import android.widget.TextView
import androidx.activity.viewModels
import co.africanwolf.hiddenwords.FullscreenActivity
import co.africanwolf.hiddenwords.R
import co.africanwolf.hiddenwords.SoundPlayer
import co.africanwolf.hiddenwords.commons.DurationFormatter
import co.africanwolf.hiddenwords.commons.Util
import co.africanwolf.hiddenwords.commons.viewBinding
import co.africanwolf.hiddenwords.custom.LetterBoard.OnLetterSelectionListener
import co.africanwolf.hiddenwords.custom.StreakView.StreakLine
import co.africanwolf.hiddenwords.data.entity.Game
import co.africanwolf.hiddenwords.data.entity.GameType
import co.africanwolf.hiddenwords.databinding.ActivityGamePlayBinding
import co.africanwolf.hiddenwords.gameover.GameoverDialog
import co.africanwolf.hiddenwords.gameover.GameoverDialog.Companion.newInstance
import co.africanwolf.hiddenwords.gameover.GameoverDialog.GameOverDialogInputListener
import co.africanwolf.hiddenwords.gameplay.GamePlayViewModel.AnswerResult
import co.africanwolf.hiddenwords.gameplay.GamePlayViewModel.GameState
import co.africanwolf.hiddenwords.gameplay.GamePlayViewModel.Generating
import co.africanwolf.hiddenwords.gameplay.GamePlayViewModel.Paused
import co.africanwolf.hiddenwords.gameplay.GamePlayViewModel.Playing
import co.africanwolf.hiddenwords.gameplay.GamePlayViewModel.Update
import co.africanwolf.hiddenwords.mainmenu.MainMenuActivity
import co.africanwolf.hiddenwords.model.GameData
import co.africanwolf.hiddenwords.model.UsedWord
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.serialization.json.Json
import javax.inject.Inject

@AndroidEntryPoint
class GamePlayActivity : FullscreenActivity(), GameOverDialogInputListener {

    @Inject
    lateinit var mSoundPlayer: SoundPlayer

    private val mViewModel: GamePlayViewModel by viewModels()

    private val binding by viewBinding(ActivityGamePlayBinding::inflate)
    private var mLetterAdapter: ArrayLetterGridDataAdapter? = null
    private lateinit var game: Game

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        binding.letterBoard.streakView.setEnableOverrideStreakLineColor(preferences.grayscale())
        binding.letterBoard.streakView.setOverrideStreakLineColor(R.color.gray)
        binding.letterBoard.setOnLetterSelectionListener(object : OnLetterSelectionListener {
            override fun onSelectionBegin(streakLine: StreakLine, str: String) {
                streakLine.color = Util.getRandomColorWithAlpha(170)
                binding.textSelection.visibility = View.VISIBLE
                binding.textSelection.text = str
            }

            override fun onSelectionDrag(streakLine: StreakLine, str: String) {
                if (str.isEmpty()) {
                    binding.textSelection.text = "..."
                } else {
                    binding.textSelection.text = str
                }
            }

            override fun onSelectionEnd(streakLine: StreakLine, str: String) {
                mViewModel.answerWord(
                    str,
                    STREAK_LINE_MAPPER.revMap(streakLine),
                    preferences.reverseMatching()
                )
                binding.textSelection.text = ""
            }
        })

        mViewModel.onTimer.observe(this) { duration: Long -> showDuration(duration) }
        mViewModel.onGameState.observe(this) { gameState: GameState ->
            onGameStateChanged(
                gameState
            )
        }
        mViewModel.onAnswerResult.observe(this) { answerResult: AnswerResult ->
            onAnswerResult(
                answerResult
            )
        }
        val extras = intent.extras
        if (extras != null) {
            if (extras.containsKey(EXTRA_GAME_ROUND_ID)) {
                val gid = extras.getInt(EXTRA_GAME_ROUND_ID)
                mViewModel.loadGameRound(gid)
            } else {
                val rowCount = extras.getInt(EXTRA_ROW_COUNT)
                val colCount = extras.getInt(EXTRA_COL_COUNT)
                val gameData = extras.getString(EXTRA_GAME_DATA)
                game = gameData?.let { Json.decodeFromString(Game.serializer(), it) }!!
                mViewModel.generateNewGameRound(rowCount, colCount, game)
                updateUI()
            }
        }
        if (!preferences.showGridLine()) {
            binding.letterBoard.gridLineBackground.visibility = View.INVISIBLE
        } else {
            binding.letterBoard.gridLineBackground.visibility = View.VISIBLE
        }
        binding.letterBoard.streakView.isSnapToGrid = preferences.snapToGrid
    }

    private fun updateUI() {
        if (game.type == GameType.BLIND_FOLD) {
            binding.flowLayout.visibility = GONE
        }

        binding.gameType.text = game.type.type
    }

    override fun onStart() {
        super.onStart()
        mViewModel.resumeGame()
    }

    override fun onStop() {
        super.onStop()
        mViewModel.pauseGame()
    }

    override fun onDestroy() {
        super.onDestroy()
        mViewModel.stopGame()
    }

    private fun onAnswerResult(answerResult: AnswerResult) {
        if (answerResult.correct) {
            val textView = findUsedWordTextViewByUsedWordId(answerResult.usedWordId)
            if (textView != null) {
                val uw = textView.tag as UsedWord
                if (preferences.grayscale()) {
                    uw.answerLine.color = R.color.gray
                }
                textView.setBackgroundColor(uw.answerLine.color)
                textView.text = uw.string
                textView.setTextColor(Color.WHITE)
                textView.paintFlags = textView.paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
                val anim = AnimatorInflater.loadAnimator(this, R.animator.zoom_in_out)
                anim.setTarget(textView)
                anim.start()
            }
            mSoundPlayer.play(SoundPlayer.Sound.Correct)
        } else {
            binding.letterBoard.popStreakLine()
            mSoundPlayer.play(SoundPlayer.Sound.Wrong)
        }
    }

    private fun onGameStateChanged(gameState: GameState) {
        showLoading(false, null)
        if (gameState is Generating) {
            val state = gameState
            val text = "Generating " + state.rowCount + "x" + state.colCount + " grid"
            showLoading(true, text)
            refresh()
        } else if (gameState is GamePlayViewModel.Finished) {
            showFinishGame(gameState.mGameData.id)
        } else if (gameState is Paused) {
        } else if (gameState is Playing) {
            onGameRoundLoaded(gameState.mGameData)
        } else if (gameState is Update) {
            showAnsweredWordsCount(gameState.mGameData)
        }
    }

    private fun refresh() {
        binding.flowLayout.removeAllViews()
        if (mLetterAdapter != null) {
            mLetterAdapter!!.clear()
        }
        binding.letterBoard.removeAllStreakLine()
    }

    private fun onGameRoundLoaded(gameData: GameData) {
        if (gameData.isFinished) {
            setGameAsAlreadyFinished()
        }
        showLetterGrid(gameData.grid.array)
        showDuration(gameData.duration)
        showUsedWords(gameData.usedWords)
        showWordsCount(gameData.usedWords.size)
        doneLoadingContent()
    }

    private fun tryScale() {
        val metrics = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(metrics)
        val boardWidth = binding.letterBoard.width
        val screenWidth = metrics.widthPixels
        if (preferences.autoScaleGrid() || boardWidth > screenWidth) {
            val scale = screenWidth.toFloat() / boardWidth.toFloat()
            binding.letterBoard.scale(scale, scale)
        }
    }

    private fun doneLoadingContent() {
        // call tryScale() on the next render frame
        Handler().postDelayed({ tryScale() }, 100)
    }

    private fun showLoading(enable: Boolean, text: String?) {
        if (enable) {
            binding.loading.visibility = View.VISIBLE
            binding.loadingText.visibility = View.VISIBLE
            binding.contentLayout.visibility = View.GONE
            binding.loadingText.text = text
        } else {
            binding.loading.visibility = View.GONE
            binding.loadingText.visibility = View.GONE
            binding.contentLayout.visibility = View.VISIBLE
        }
    }

    private fun showLetterGrid(grid: Array<CharArray>) {
        if (mLetterAdapter == null) {
            mLetterAdapter = ArrayLetterGridDataAdapter(grid)
            binding.letterBoard.dataAdapter = mLetterAdapter
        } else {
            mLetterAdapter!!.grid = grid
        }
    }

    private fun showDuration(duration: Long) {
        binding.textDuration.text = DurationFormatter.fromInteger(duration)
    }

    private fun showUsedWords(usedWords: List<UsedWord>) {
        for (uw in usedWords) {
            binding.flowLayout.addView(createUsedWordTextView(uw))
        }
    }

    private fun showAnsweredWordsCount(gameData: GameData) {
        binding.foundWordsCount.text = getString(
            R.string.word_count, gameData.answeredWordsCount,
            gameData.usedWords.size
        )
    }

    private fun showWordsCount(count: Int) {
        binding.foundWordsCount.text = getString(R.string.word_count, 0, count)
    }

    private fun showFinishGame(gameId: Int) {
        /*    Intent intent = new Intent(this, GameOverActivity.class);
        intent.putExtra(GameOverActivity.EXTRA_GAME_ROUND_ID, gameId);
        startActivity(intent);
        finish();*/
        val fm = supportFragmentManager
        val editNameDialogFragment = newInstance("Some Title")
        editNameDialogFragment.show(fm, "fragment_edit_name")
    }

    private fun setGameAsAlreadyFinished() {
        binding.letterBoard.streakView.isInteractive = false
    }

    //
    private fun createUsedWordTextView(uw: UsedWord): TextView {
        val tv = TextView(this)
        tv.setPadding(10, 5, 10, 5)
        if (uw.isAnswered) {
            if (preferences.grayscale()) {
                uw.answerLine.color = R.color.gray
            }
            tv.setBackgroundColor(uw.answerLine.color)
            tv.text = uw.string
            tv.paintFlags = tv.paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
            binding.letterBoard.addStreakLine(STREAK_LINE_MAPPER.map(uw.answerLine))
        } else {
            var str = uw.string
            if (uw.isMystery) {
                var revealCount = uw.revealCount
                val uwString = uw.string
                str = ""
                for (element in uwString) {
                    if (revealCount > 0) {
                        str += element
                        revealCount--
                    } else {
                        str += " _"
                    }
                }
            }
            tv.text = "$str "
        }
        tv.setTextColor(Color.WHITE)
        tv.setTextSize(TypedValue.COMPLEX_UNIT_SP, 20f)
        tv.tag = uw
        return tv
    }

    private fun findUsedWordTextViewByUsedWordId(usedWordId: Int): TextView? {
        for (i in 0 until binding.flowLayout.childCount) {
            val tv = binding.flowLayout.getChildAt(i) as TextView
            val uw = tv.tag as UsedWord
            if (uw != null && uw.id == usedWordId) {
                return tv
            }
        }
        return null
    }

    override fun sendInput(input: String?) {
        if (input == GameoverDialog.ACTION_MAIN_MENU) {
            val intent = Intent(this, MainMenuActivity::class.java)
            startActivity(intent)
            finish()
        } else {
            newGame()
        }
    }

    private fun newGame() {
        val extras = intent.extras
        val rowCount = extras!!.getInt(EXTRA_ROW_COUNT)
        val colCount = extras.getInt(EXTRA_COL_COUNT)
        mViewModel.generateNewGameRound(rowCount, colCount, game)
    }

    companion object {
        const val EXTRA_GAME_ROUND_ID = "co.africanwolf.hiddenwords.gameplay.GamePlayActivity.ID"
        const val EXTRA_ROW_COUNT = "co.africanwolf.hiddenwords.gameplay.GamePlayActivity.ROW"
        const val EXTRA_COL_COUNT = "co.africanwolf.hiddenwords.gameplay.GamePlayActivity.COL"
        const val EXTRA_GAME_DATA =
            "co.africanwolf.hiddenwords.gameplay.GamePlayActivity.GAME_DATA"
        private val STREAK_LINE_MAPPER = StreakLineMapper()
    }
}
